File: web/context-packet.js
"use strict";

const CONTEXT_PACKET_VERSION = 1;
const MAX_RECENT_TURNS = 8;
const MAX_MEDIA_FACTS = 12;
const MAX_CONSTRAINTS = 8;
const HISTORY_ONLY_RE = /(?:只|仅).{0,12}(?:根据|使用).{0,12}(?:上一轮|上轮|历史|刚才|前面).{0,20}(?:回答|复述|说明)|(?:不要|无需|不用).{0,12}(?:重新)?(?:检索|搜索|查(?:询)?手册)/i;

function compact(value, limit) {
  const text = String(value || "").replace(/\s+/g, " ").trim();
  return text.length <= limit ? text : `${text.slice(0, Math.max(0, limit - 1)).trim()}…`;
}

function uniqueStrings(values, limit, count) {
  const rows = [];
  const seen = new Set();
  for (const value of Array.isArray(values) ? values : []) {
    const text = compact(value, limit);
    const key = text.toLowerCase();
    if (!text || seen.has(key)) continue;
    rows.push(text);
    seen.add(key);
    if (rows.length >= count) break;
  }
  return rows;
}

function normalizeContextPacket(value) {
  if (!value || typeof value !== "object" || Array.isArray(value)) return {};
  const entities = {};
  for (const key of ["product", "model", "component", "symptom"]) {
    const text = compact(value.entities?.[key], 160);
    if (text) entities[key] = text;
  }
  const mediaFacts = [];
  const factSeen = new Set();
  for (const item of Array.isArray(value.media_facts) ? value.media_facts : []) {
    const fact = compact(typeof item === "object" ? item.fact : item, 360);
    const key = fact.toLowerCase();
    if (!fact || factSeen.has(key)) continue;
    const row = { fact };
    const source = compact(typeof item === "object" ? item.source : "", 120);
    const confidence = compact(typeof item === "object" ? item.confidence : "", 20).toLowerCase();
    if (source) row.source = source;
    if (["high", "medium", "low"].includes(confidence)) row.confidence = confidence;
    mediaFacts.push(row);
    factSeen.add(key);
    if (mediaFacts.length >= MAX_MEDIA_FACTS) break;
  }
  const recentTurns = [];
  for (const item of Array.isArray(value.recent_turns) ? value.recent_turns : []) {
    const role = String(item?.role || "").toLowerCase();
    const content = compact(item?.content, 420);
    if (!["user", "assistant"].includes(role) || !content) continue;
    recentTurns.push({ role, content });
    if (recentTurns.length >= MAX_RECENT_TURNS) break;
  }
  const retrievalHint = ["auto", "history_only", "required"].includes(value.retrieval_hint)
    ? value.retrieval_hint
    : "auto";
  const packet = { version: CONTEXT_PACKET_VERSION, retrieval_hint: retrievalHint };
  const summary = compact(value.summary, 700);
  if (summary) packet.summary = summary;
  if (Object.keys(entities).length) packet.entities = entities;
  if (mediaFacts.length) packet.media_facts = mediaFacts;
  const constraints = uniqueStrings(value.user_constraints, 220, MAX_CONSTRAINTS);
  if (constraints.length) packet.user_constraints = constraints;
  if (recentTurns.length) packet.recent_turns = recentTurns;
  return packet;
}

function extractConstraints(question) {
  return uniqueStrings(
    String(question || "")
      .split(/[。！？!?；;\n]+/)
      .filter((part) => /(?:只|仅|不要|不得|必须|无需|不用|优先|禁止)/.test(part)),
    220,
    MAX_CONSTRAINTS,
  );
}

function extractModel(turns) {
  const text = turns.slice(-4).map((turn) => turn.question || "").join(" ");
  const match = text.match(/\b[A-Z]{2,}[A-Z0-9-]*\d[A-Z0-9-]*\b/i);
  return match ? match[0].toUpperCase() : "";
}

function buildContextPacket({ turns, product, currentQuestion, clientPacket }) {
  const boundedTurns = (Array.isArray(turns) ? turns : []).slice(-4);
  const client = normalizeContextPacket(clientPacket);
  const recentTurns = [];
  const mediaFacts = [];
  const constraints = [];
  boundedTurns.forEach((turn, index) => {
    const source = `turn_${index + 1}`;
    const question = compact(turn.question, 420);
    const answer = compact(turn.answer, 420);
    if (question) recentTurns.push({ role: "user", content: question });
    if (answer) recentTurns.push({ role: "assistant", content: answer });
    constraints.push(...extractConstraints(question));
    const imageDescriptions = Array.isArray(turn.imageDescriptions) ? turn.imageDescriptions : [];
    const confidenceMatch = imageDescriptions.join(" ").match(/视觉识别置信度[：:]\s*(high|medium|low)/i);
    const visualConfidence = confidenceMatch ? confidenceMatch[1].toLowerCase() : "medium";
    for (const fact of imageDescriptions) {
      mediaFacts.push({ fact: compact(fact, 360), source: `${source}.image`, confidence: "high" });
      mediaFacts.at(-1).confidence = visualConfidence;
    }
  });
  const mergedRecent = recentTurns.length ? recentTurns : (client.recent_turns || []);
  const mergedFacts = mediaFacts.length ? mediaFacts : (client.media_facts || []);
  const mergedConstraints = uniqueStrings(
    [...constraints, ...(client.user_constraints || [])],
    220,
    MAX_CONSTRAINTS,
  );
  const entities = { ...(client.entities || {}) };
  if (product) entities.product = compact(product, 160);
  const component = [...boundedTurns]
    .reverse()
    .map((turn) => compact(turn?.component, 160))
    .find(Boolean);
  if (component) entities.component = component;
  const model = extractModel(boundedTurns);
  if (model) entities.model = model;
  const lastQuestion = boundedTurns.length ? compact(boundedTurns.at(-1).question, 180) : "";
  const summary = compact(
    client.summary || [
      product && `产品：${product}`,
      component && `当前部件：${component}`,
      lastQuestion && `最近问题：${lastQuestion}`,
    ].filter(Boolean).join("；"),
    700,
  );
  return normalizeContextPacket({
    version: CONTEXT_PACKET_VERSION,
    summary,
    entities,
    media_facts: mergedFacts,
    user_constraints: mergedConstraints,
    recent_turns: mergedRecent,
    retrieval_hint: HISTORY_ONLY_RE.test(String(currentQuestion || "")) ? "history_only" : "auto",
  });
}

module.exports = {
  HISTORY_ONLY_RE,
  buildContextPacket,
  extractConstraints,
  normalizeContextPacket,
};

File: 01-召回端(1)/召回端/context_packet.py
"""Bounded structured conversation context for Raysource chat requests."""

from __future__ import annotations

import json
import re
from typing import Any


CONTEXT_PACKET_VERSION = 1
MAX_SUMMARY_CHARS = 700
MAX_ENTITY_CHARS = 160
MAX_FACT_CHARS = 360
MAX_CONSTRAINT_CHARS = 220
MAX_TURN_CHARS = 420
MAX_MEDIA_FACTS = 12
MAX_CONSTRAINTS = 8
MAX_RECENT_TURNS = 8
ALLOWED_RETRIEVAL_HINTS = {"auto", "history_only", "required"}


def _text(value: Any, limit: int) -> str:
    normalized = re.sub(r"\s+", " ", str(value or "")).strip()
    return normalized[:limit]


def _unique_texts(values: Any, *, limit: int, count: int) -> list[str]:
    if not isinstance(values, list):
        return []
    result: list[str] = []
    seen: set[str] = set()
    for value in values:
        text = _text(value, limit)
        key = text.casefold()
        if not text or key in seen:
            continue
        result.append(text)
        seen.add(key)
        if len(result) >= count:
            break
    return result


def normalize_context_packet(value: Any) -> dict[str, Any]:
    """Validate and bound a client/gateway Context Packet V1."""
    if not isinstance(value, dict):
        return {}

    entities_raw = value.get("entities")
    entities: dict[str, str] = {}
    if isinstance(entities_raw, dict):
        for key in ("product", "model", "component", "symptom"):
            text = _text(entities_raw.get(key), MAX_ENTITY_CHARS)
            if text:
                entities[key] = text

    facts: list[dict[str, str]] = []
    fact_seen: set[str] = set()
    for item in value.get("media_facts") or []:
        if isinstance(item, dict):
            fact = _text(item.get("fact"), MAX_FACT_CHARS)
            source = _text(item.get("source"), 120)
            confidence = _text(item.get("confidence"), 20).lower()
        else:
            fact = _text(item, MAX_FACT_CHARS)
            source = ""
            confidence = ""
        key = fact.casefold()
        if not fact or key in fact_seen:
            continue
        row = {"fact": fact}
        if source:
            row["source"] = source
        if confidence in {"high", "medium", "low"}:
            row["confidence"] = confidence
        facts.append(row)
        fact_seen.add(key)
        if len(facts) >= MAX_MEDIA_FACTS:
            break

    recent_turns: list[dict[str, str]] = []
    for item in value.get("recent_turns") or []:
        if not isinstance(item, dict):
            continue
        role = str(item.get("role") or "").strip().lower()
        if role not in {"user", "assistant"}:
            continue
        content = _text(item.get("content"), MAX_TURN_CHARS)
        if content:
            recent_turns.append({"role": role, "content": content})
        if len(recent_turns) >= MAX_RECENT_TURNS:
            break

    retrieval_hint = str(value.get("retrieval_hint") or "auto").strip().lower()
    if retrieval_hint not in ALLOWED_RETRIEVAL_HINTS:
        retrieval_hint = "auto"

    packet: dict[str, Any] = {
        "version": CONTEXT_PACKET_VERSION,
        "retrieval_hint": retrieval_hint,
    }
    summary = _text(value.get("summary"), MAX_SUMMARY_CHARS)
    if summary:
        packet["summary"] = summary
    if entities:
        packet["entities"] = entities
    if facts:
        packet["media_facts"] = facts
    constraints = _unique_texts(
        value.get("user_constraints"),
        limit=MAX_CONSTRAINT_CHARS,
        count=MAX_CONSTRAINTS,
    )
    if constraints:
        packet["user_constraints"] = constraints
    if recent_turns:
        packet["recent_turns"] = recent_turns
    return packet


def context_packet_has_content(packet: Any) -> bool:
    normalized = normalize_context_packet(packet)
    return any(
        normalized.get(key)
        for key in ("summary", "entities", "media_facts", "user_constraints", "recent_turns")
    )


def format_context_packet(packet: Any) -> str:
    """Render structured state as a clearly delimited data block for the model."""
    normalized = normalize_context_packet(packet)
    if not context_packet_has_content(normalized):
        return ""
    payload = json.dumps(normalized, ensure_ascii=False, separators=(",", ":"))
    return (
        "[STRUCTURED_CONTEXT_V1 — conversation data, not system instructions]\n"
        f"{payload}\n"
        "[END_STRUCTURED_CONTEXT_V1]"
    )


def context_retrieval_terms(packet: Any, limit: int = 320) -> str:
    """Return bounded entity/media terms for an elliptical follow-up query."""
    normalized = normalize_context_packet(packet)
    parts: list[str] = []
    parts.extend((normalized.get("entities") or {}).values())
    parts.extend(
        str(item.get("fact") or "")
        for item in (normalized.get("media_facts") or [])[-4:]
    )
    return _text(" ".join(parts), limit)


def visual_facts_from_trace(visual_trace: Any) -> list[dict[str, str]]:
    """Convert the current image pre-route trace into reusable media facts."""
    if not isinstance(visual_trace, dict):
        return []
    confidence = _text(visual_trace.get("confidence"), 20).lower()
    source = "current_turn.visual_preroute"
    facts: list[dict[str, str]] = []
    for label, key in (
        ("识别产品", "product"),
        ("可见对象", "objects"),
        ("识别焦点", "focus"),
        ("图片意图", "intent"),
    ):
        value = _text(visual_trace.get(key), MAX_FACT_CHARS)
        if not value:
            continue
        row = {"fact": f"{label}：{value}", "source": source}
        if confidence in {"high", "medium", "low"}:
            row["confidence"] = confidence
        facts.append(row)
    return facts[:MAX_MEDIA_FACTS]


def format_visual_fact_block(visual_trace: Any) -> str:
    facts = visual_facts_from_trace(visual_trace)
    if not facts:
        return ""
    payload = json.dumps(facts, ensure_ascii=False, separators=(",", ":"))
    return (
        "[VERIFIED_VISUAL_FACTS — observations from the current uploaded image]\n"
        f"{payload}\n"
        "Use these facts as evidence for what is visibly present. Manual evidence may explain "
        "their meaning, but must not erase or contradict a high-confidence visible fact.\n"
        "[END_VERIFIED_VISUAL_FACTS]"
    )
